pallets/remote-proxy/src/lib.rs

```diff
@@ -0,0 +1,496 @@
+		/// - `force_proxy_type`: Specify the exact proxy type to be used and checked for this call.
+		/// - `call`: The call to be made by the `real` account.
+		#[pallet::call_index(2)]
+		#[pallet::weight({
+			let di = call.get_dispatch_info();
+			(WeightInfoOf::<T, I>::remote_proxy_with_registered_proof()
+				// AccountData for inner call origin accountdata.
+				.saturating_add(T::DbWeight::get().reads_writes(1, 1))
+				.saturating_add(di.weight),
+			di.class)
+		})]
+		pub fn remote_proxy_with_registered_proof(
+			origin: OriginFor<T>,
+			real: AccountIdLookupOf<T>,
+			force_proxy_type: Option<T::ProxyType>,
+			call: Box<<T as pallet_proxy::Config>::RuntimeCall>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let real = T::Lookup::lookup(real)?;
+
+			let proof = dispatch_context::with_context::<
+				RemoteProxyContext<RemoteBlockNumberOf<T, I>>,
+				_,
+			>(|context| context.or_default().proofs.pop())
+			.flatten()
+			.ok_or(Error::<T, I>::ProxyProofNotRegistered)?;
+
+			Self::do_remote_proxy(who, real, force_proxy_type, *call, proof)
+		}
+	}
+
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		fn do_remote_proxy(
+			who: T::AccountId,
+			real: T::AccountId,
+			force_proxy_type: Option<T::ProxyType>,
+			call: <T as pallet_proxy::Config>::RuntimeCall,
+			proof: RemoteProxyProof<RemoteBlockNumberOf<T, I>>,
+		) -> DispatchResult {
+			let Some(real_remote) = T::RemoteProxy::local_to_remote_account_id(&real) else {
+				return Err(Error::<T, I>::CouldNotConvertLocalToRemoteAccountId.into());
+			};
+
+			let def = match proof {
+				RemoteProxyProof::RelayChain { proof, block } => {
+					let roots = BlockToRoot::<T, I>::get();
+
+					let Ok(storage_root) = roots
+						.binary_search_by(|(b, _)| b.cmp(&block))
+						.map(|pos| roots[pos].1.clone())
+					else {
+						return Err(Error::<T, I>::UnknownProofAnchorBlock.into());
+					};
+
+					let key = T::RemoteProxy::proxy_definition_storage_key(&real_remote);
+
+					let db =
+						sp_trie::StorageProof::new(proof).into_memory_db::<RemoteHasherOf<T, I>>();
+					let value = sp_trie::read_trie_value::<sp_trie::LayoutV1<_>, _>(
+						&db,
+						&storage_root,
+						&key,
+						None,
+						None,
+					)
+					.ok()
+					.flatten()
+					.ok_or(Error::<T, I>::InvalidProof)?;
+
+					let proxy_definitions = alloc::vec::Vec::<
+						ProxyDefinition<
+							RemoteAccountIdOf<T, I>,
+							RemoteProxyTypeOf<T, I>,
+							RemoteBlockNumberOf<T, I>,
+						>,
+					>::decode(&mut &value[..])
+					.map_err(|_| Error::<T, I>::ProxyDefinitionDecodingFailed)?;
+
+					let f = |x: &ProxyDefinition<
+						T::AccountId,
+						T::ProxyType,
+						BlockNumberFor<T>,
+					>|
+					 -> bool {
+						x.delegate == who &&
+							force_proxy_type.as_ref().map_or(true, |y| &x.proxy_type == y)
+					};
+
+					proxy_definitions
+						.into_iter()
+						.filter_map(T::RemoteProxy::remote_to_local_proxy_defintion)
+						.find(f)
```

Comment:
I guess in the case that there are multiple proxies and no force type, it will just pick the first one?  
Should not produce issues i think, just an observation.